### PR TITLE
feat: Add functionality to export doc strings on types

### PR DESCRIPTION
**Pull request by timephy**
# Overview

I added functionality (and tests) to export doc strings on exported types.

# Example

```rust
/// Doc comment.
/// Supports new lines.
///
/// Testing
#[derive(TS)]
#[ts(export_to = "tests-out/docs/")]
struct A {
    name: String,
}
```

is exported as

```ts
/**
 * Doc comment.
 * Supports new lines.
 *
 * Testing
 */
export interface A { name: string, }
```

# How

This is achieved by parsing the `Vec<Attribute>` of an the exported type `Item` by filtering for `"doc"` , and then passing that data down to the `export` layer where `generate_decl` prepends the doc string before the export definition in TypeScript.

# Comments

- The only thing I am unsure about is how this interacts with the `type_def` function which is used both for exported types and field types.
- Only works with "normal doc comments", not with `include_str` doc comments or other "doc formats".

# Related

#52
#102

### CODE REVIEW

**Comment by bennyhodl** [comment]
Concept ACK. Would love to see this merged

https://github.com/Aleph-Alpha/ts-rs/issues/188

**Comment by escritorio-gustavo** [comment]
Fixes #188 
Just so GitHub tracks the issue with the PR

**Comment by NyxCode** [comment]
Thanks for the PR! 
The general structure of the PR looks good, though there are some minor problems I'd like to resolve before merging this.

**Comment by escritorio-gustavo** [comment]
Hey @NyxCode! What are those minor problems? I can take a look when I have time if you want

**Comment by NyxCode** [comment]
> Hey @NyxCode! What are those minor problems? I can take a look when I have time if you want

oh, seems like i didnt submit the review, sorry.

**Review by NyxCode — CHANGES_REQUESTED** [review]

**Comment by escritorio-gustavo** [comment]
Well this is annoying, locally editing this PR gets me 131 commits behind main, so tests still expect `interface`, but on CI they expect `type`

**Comment by escritorio-gustavo** [comment]
@timephy, can you merge main into your fork? I tried to do it but it just turns into a PR to your fork

**Comment by escritorio-gustavo** [comment]
Nevermind, figured out how to do it!
Sorry, still new to managing PRs

File: macros/src/types/mod.rs [hunk 1]
```diff
@@ -14,26 +18,28 @@ pub(crate) use r#enum::r#enum_def;
...
 
 fn type_def(
     attr: &StructAttr,
+    attrs: &Vec<Attribute>,
     ident: &Ident,
     fields: &Fields,
     generics: &Generics,
 ) -> Result<DerivedTS> {
     let name = attr.rename.clone().unwrap_or_else(|| to_ts_ident(ident));
+    let docs = utils::get_docs_from_attributes(attrs);
```
[the remark this answers is not in the document]
Comment (escritorio-gustavo) [reply macros/src/types/mod.rs:32]: Also needs to be done for `EnumAttr`

[the remark this answers is not in the document]
Comment (escritorio-gustavo) [reply macros/src/types/mod.rs]: Also renamed the function to parse_docs

File: macros/src/utils.rs [hunk 2]
```diff
@@ -116,6 +116,24 @@ pub fn parse_serde_attrs<'a, A: TryFrom<&'a Attribute, Error = Error>>(
...
+            _ => None,
+        })
+        .map(|attr| match attr.value {
+            Expr::Lit(ExprLit {
+                lit: Lit::Str(ref str),
+                ..
+            }) => Ok(str.value()),
+            _ => syn_err!(attr.span(); "doc attribute with non literal expression found"),
+        })
+        .collect::<Result<Vec<_>>>()
```
Comment (escritorio-gustavo) [inline macros/src/utils.rs:134]: Managed to get rid of those `panic!`s

File: macros/src/utils.rs [hunk 3]
```diff
@@ -116,6 +116,24 @@ pub fn parse_serde_attrs<'a, A: TryFrom<&'a Attribute, Error = Error>>(
         .into_iter()
 }
 
+/// Return a vector of all lines of doc comments in the given vector of attributes.
+pub fn parse_docs(attrs: &[Attribute]) -> Result<Vec<String>> {
```
Comment (escritorio-gustavo) [inline macros/src/utils.rs:120]: Always prefer `&[T]` over `&Vec<T>`

File: ts-rs/src/export.rs [hunk 4]
```diff
@@ -249,3 +256,17 @@ where
...
     }
 }
+
+/// Returns an unindented docstring that has a newline at the end if it has content.
+fn format_docs(docs: &str) -> String {
+    match docs.is_empty() {
+        true => "".to_string(),
+        false => {
+            let lines = docs
+                .lines()
```
Comment (escritorio-gustavo) [inline ts-rs/src/export.rs:266]: Interestingly, changing docs from `Vec<String>` to `&str` only changed this line (`.iter()` to `.lines()`) in `format_docs`

**Review by escritorio-gustavo — COMMENTED** [review]
These are the notable changes I made on top of what @NyxCode asked

**Comment by timephy** [comment]
Hi!

First of all thank your for refactoring and fixing problems!

I have changed the code to still generate a `const DOCS = None` field, even if no docs are set on the type.

I use the `.docs()` or now `.DOCS` in code of mine, and I rely on its presence, therefore if the field is simply missing (instead of being None) this would lead to errors.

**Comment by timephy** [comment]
Hi,

First of all, thanks for refactoring the code and solving problems!
Especially that you removed `docs` from all function calls, and moved it into `Attr` structs. - This bothered me as well, but I was not sure about how to solve it better at the time.

Note: I made a commit I then realized was unnecessary, because the `TS` trait already implemented a fallback for the `DOCS` field. Therefore I removed those changes again.

---

I believe it is possible to also add docs for struct fields and enum variants... (#188)
Do we want to implement and include this in this PR or implement it somewhen else in another?

**Comment by timephy** [comment]
Okay, I tried to implement docs for fields+variants as well..

But there were several complications that we need to resolve, I'd like to ask for your help doing that / making the right decisions. I have committed these changes to a separate branch and opened a new PR #228 for that.

I suggest implementing this PR first / implementing the docs for fields/variants in that new PR.

Thanks again for the help/coop!

**Comment by escritorio-gustavo** [comment]
> I believe it is possible to also add docs for struct fields and enum variants

I think struct fields shouldn't be **too** dificult, but we have to make sure to handle `#[ts(flatten)]` correctly, which always complicates things

Enum variants are a whole other beast though. They are represented as a type union of the different variants and there is no convenient way to add docs for specific type union members in TS.

We could implement #86 and document each variant type, but at the end of the day, we'd do `type Enum = VariantA | VariantB`, and when using `Enum` the docs for `VariantA` and `VariantB` would not show up anywhere

**Comment by escritorio-gustavo** [comment]
> Do we want to implement and include this in this PR or implement it somewhen else in another?

Definetly another PR, creating #228 was a good call

> I suggest implementing this PR first / implementing the docs for fields/variants in that new PR.

Agreed, since all the changes requested have already been made, I will merge this PR now

**Pull request merged**